services/notifications/pkg/service/service.go

```diff
@@ -87,52 +97,101 @@ func (s eventsNotifier) Run() error {
+	opaqueID string
+	subject  string
+	msg      string
 }
 
-func (s eventsNotifier) send(ctx context.Context, u *user.UserId, g *group.GroupId, msg, subj, sender string) error {
-	if u != nil {
-		return s.channel.SendMessage(ctx, []string{u.GetOpaqueId()}, msg, subj, sender)
+func (s eventsNotifier) render(ctx context.Context, template email.MessageTemplate,
+	granteeFieldName string, fields map[string]interface{}, granteeList []*user.UserId) ([]*recipient, error) {
+	// Render the Email Template for each user
+	recipientList := make([]*recipient, len(granteeList))
+	for i, userID := range granteeList {
+		locale, err := s.getUserLang(ctx, userID)
+		if err != nil {
+			return nil, err
+		}
+		grantee, err := s.getUserName(ctx, userID)
+		if err != nil {
+			return nil, err
+		}
+		fields[granteeFieldName] = grantee
 
+		subj, msg, err := email.RenderEmailTemplate(template, locale, s.emailTemplatePath, s.translationPath, fields)
+		if err != nil {
+			return nil, err
+		}
+		recipientList[i] = &recipient{opaqueID: userID.GetOpaqueId(), subject: subj, msg: msg}
 	}
+	return recipientList, nil
+}
 
-	if g != nil {
-		return s.channel.SendMessageToGroup(ctx, g, msg, subj, sender)
+func (s eventsNotifier) send(ctx context.Context, recipientList []*recipient, sender string) {
+	for _, r := range recipientList {
+		err := s.channel.SendMessage(ctx, []string{r.opaqueID}, r.msg, r.subject, sender)
+		if err != nil {
+			s.logger.Error().Err(err).Str("event", "SendEmail").Msg("failed to send a message")
+		}
 	}
-
-	return nil
 }
 
-func (s eventsNotifier) getGranteeName(ctx context.Context, u *user.UserId, g *group.GroupId) (string, error) {
+func (s eventsNotifier) getGranteeList(ctx context.Context, owner, u *user.UserId, g *group.GroupId) ([]*user.UserId, error) {
 	switch {
 	case u != nil:
-		r, err := s.gwClient.GetUser(ctx, &user.GetUserRequest{UserId: u})
-		if err != nil {
-			return "", err
-		}
-
-		if r.Status.Code != rpc.Code_CODE_OK {
-			return "", fmt.Errorf("unexpected status code from gateway client: %d", r.GetStatus().GetCode())
-		}
-
-		return r.GetUser().GetDisplayName(), nil
+		return []*user.UserId{u}, nil
 	case g != nil:
-		r, err := s.gwClient.GetGroup(ctx, &group.GetGroupRequest{GroupId: g})
+		res, err := s.gwClient.GetGroup(ctx, &group.GetGroupRequest{GroupId: g})
 		if err != nil {
-			return "", err
+			return nil, err
 		}
-
-		if r.GetStatus().GetCode() != rpc.Code_CODE_OK {
-			return "", fmt.Errorf("unexpected status code from gateway client: %d", r.GetStatus().GetCode())
+		if res.Status.Code != rpc.Code_CODE_OK {
+			return nil, errors.New("could not get group")
 		}
-
-		return r.GetGroup().GetDisplayName(), nil
+		for i, userID := range res.GetGroup().GetMembers() {
+			// remove an owner from a list
+			if userID.GetOpaqueId() == owner.GetOpaqueId() {
+				res.Group.Members[i] = res.Group.Members[len(res.Group.Members)-1]
+				return res.Group.Members[:len(res.Group.Members)-1], nil
+			}
+		}
```

Comment:
Why is it necessary to remove the owner?

Reply:
To avoid send the mails like this
`Hello Roman

Roman has shared "secrets of the board" with you.

Click here to view it: files/shares/with-me`

Reply:
I can remove, if it not necessary.

Reply:
Valid point. But it doesn't work on project spaces. We need to remove the `Executant` which is present in most events

Reply:
Compare `userlog` service: https://github.com/owncloud/ocis/blob/master/services/userlog/pkg/service/service.go#L153

Reply:
The owner is an Executant in the handleSpaceShared and handleSpaceUnshared methods context

https://github.com/owncloud/ocis/blob/e0b6ca0e29c94f7aaa0f509be1e06ebf6a9e4b78/services/notifications/pkg/service/spaces.go#L16
https://github.com/owncloud/ocis/blob/e0b6ca0e29c94f7aaa0f509be1e06ebf6a9e4b78/services/notifications/pkg/service/spaces.go#L80

Reply:
Ah! Would you mind renaming `owner` and `ownerCtx` to `executant` and `executantCtx`? Then it's much clearer

Reply:
Also please update the comment. It's not the `owner` that is removed. It's the `executant`

Reply:
And keep the owner and ownerCtx for handleSpaceMembershipExpired right ?
https://github.com/owncloud/ocis/blob/e0b6ca0e29c94f7aaa0f509be1e06ebf6a9e4b78/services/notifications/pkg/service/spaces.go#L143

Reply:
done